lib/lightning_web/controllers/api/workflows_controller.ex

```diff
@@ -0,0 +1,461 @@
+      list =
+        Workflows.list_project_workflows(project_id,
+          include: [:edges, :jobs, :triggers]
+        )
+
+      json(conn, %{workflows: list, errors: []})
+    end
+  end
+
+  def create(conn, %{"project_id" => project_id} = params) do
+    with :ok <- validate_project_id(conn.body_params, project_id),
+         :ok <- authorize_write(conn, project_id),
+         {:ok, %{id: workflow_id}} <-
+           save_workflow(params, conn.assigns.current_resource) do
+      conn
+      |> put_status(:created)
+      |> json(%{id: workflow_id, errors: []})
+    end
+    |> then(&maybe_handle_error(conn, &1))
+  end
+
+  def show(conn, %{"project_id" => project_id, "id" => workflow_id}) do
+    with :ok <- authorize_read(conn, project_id),
+         {:ok, workflow} <- get_workflow(workflow_id, project_id) do
+      json(conn, %{workflow: workflow, errors: []})
+    end
+  end
+
+  def update(conn, %{"project_id" => project_id, "id" => workflow_id} = params) do
+    with :ok <- validate_project_id(conn.body_params, project_id),
+         :ok <- authorize_write(conn, project_id),
+         {:ok, workflow} <- get_workflow(workflow_id, project_id),
+         :ok <- authorize_write(conn, workflow),
+         {:ok, %{id: workflow_id}} <-
+           save_workflow(workflow, params, conn.assigns.current_resource) do
+      json(conn, %{id: workflow_id, errors: []})
+    end
+    |> then(&maybe_handle_error(conn, &1, workflow_id))
+  end
+
+  defp count_enabled_triggers(params),
+    do: params |> Map.get("triggers", []) |> Enum.count(& &1["enabled"])
+
+  defp save_workflow(%{"project_id" => project_id} = params, user) do
+    active_triggers_count = count_enabled_triggers(params)
+
+    cond do
+      active_triggers_count > 1 ->
+        {:error, :too_many_active_triggers}
+
+      :else ->
+        save_workflow(params, active_triggers_count > 0, project_id, user)
+    end
+  end
+
+  defp save_workflow(
+         %{id: workflow_id, triggers: triggers} = workflow,
+         params,
+         user
+       ) do
+    changes_triggers? = Map.has_key?(params, "triggers")
+
+    triggers_ids =
+      params
+      |> Map.get("triggers", [])
+      |> Enum.map(& &1["id"])
+
+    active_triggers_count = count_enabled_triggers(params)
+
+    has_external_reference? = has_external_reference?(params, workflow_id)
+
+    cond do
+      changes_triggers? and Enum.any?(triggers, &(&1.id not in triggers_ids)) ->
+        {:error, :cannot_replace_trigger}
+
+      active_triggers_count > 1 ->
+        {:error, :too_many_active_triggers}
+
+      has_external_reference? ->
+        {:error, :invalid_workflow_id}
+
+      :else ->
+        workflow
+        |> Workflows.change_workflow(params)
+        |> save_workflow(active_triggers_count > 0, workflow.project_id, user)
+    end
+  end
+
+  defp save_workflow(params_or_changeset, activate?, project_id, user) do
+    with :ok <- check_limit(activate?, project_id),
```

Comment:
There is a `lib/lightning/workflows/workflow_usage_limiter.ex` module with these helpers already, did you consider using it? 
If given a `changeset`, it will determine the activated ones  and call the limiter.

Reply:
Sounds good